pkgs/development/compilers/opensmalltalk-vm/default.nix

```diff
@@ -103,84 +92,102 @@ let
     };
 
-  vmsByPlatform = {
-    "aarch64-linux" = {
-      "squeak-cog-spur" = buildVM {
+  buildVM = { vmName, platforms }:
+    stdenv.mkDerivation (final:
+      let
+        pkgFunc = platforms.${stdenv.targetPlatform.system} or (_: _: { });
+        prev = baseFunc {
+          inherit vmName;
+          platforms = builtins.attrNames platforms;
+        } final;
+      in prev // pkgFunc final prev);
+
+in {
+  squeak-cog-spur = buildVM {
+    vmName = "squeak.cog.spur";
+    platforms = {
+      aarch64-linux = (final: prev: {
+        scriptName = "squeak";
         platformDir = "linux64ARMv8";
+        CFLAGS = prev.CFLAGS ++ [
+          "-DMUSL"
+          "-D_GNU_SOURCE"
+          "-DUSEEVDEV"
+          "-DCOGMTVM=0"
+          "-DDUAL_MAPPED_CODE_ZONE=1"
+        ];
+        LIBS = [ "-lrt" ];
+        configureFlags = prev.configureFlags ++ [
+          "--with-vmversion=5.0"
+          "--with-src=src/spur64.cog"
+          "--without-npsqueak"
+          "--enable-fast-bitblt"
+        ];
+      });
+      x86_64-linux = (final: prev: {
         vmName = "squeak.cog.spur";
         scriptName = "squeak";
-        configureFlagsArray = ''(
-          CFLAGS="-DNDEBUG -DDEBUGVM=0 -DMUSL -D_GNU_SOURCE -DUSEEVDEV -DCOGMTVM=0 -DDUAL_MAPPED_CODE_ZONE=1"
-          LIBS="-lrt"
-        )'';
-        configureFlags = [
+        platformDir = "linux64x64";
+        CFLAGS = prev.CFLAGS ++ [ "-DCOGMTVM=0" ];
+        configureFlags = prev.configureFlags ++ [
           "--with-vmversion=5.0"
           "--with-src=src/spur64.cog"
           "--without-npsqueak"
-          "--enable-fast-bitblt"
         ];
-      };
+      });
+    };
+  };
 
-      "squeak-stack-spur" = buildVM {
-        platformDir = "linux64ARMv8";
-        vmName = "squeak.stack.spur";
+  squeak-stack-spur = buildVM {
+    vmName = "squeak.stack.spur";
+    platforms = {
+      aarch64-linux = (final: prev: {
         scriptName = "squeak";
-        configureFlagsArray = ''(
-          CFLAGS="-DNDEBUG -DDEBUGVM=0 -DMUSL -D_GNU_SOURCE -DUSEEVDEV -D__ARM_ARCH_ISA_A64 -DARM64 -D__arm__ -D__arm64__ -D__aarch64__"
-        )'';
-        configureFlags = [
+        platformDir = "linux64ARMv8";
+        CFLAGS = prev.CFLAGS ++ [
+          "-DMUSL"
+          "-D_GNU_SOURCE"
+          "-DUSEEVDEV"
+          "-D__ARM_ARCH_ISA_A64"
+          "-DARM64"
+          "-D__arm__"
+          "-D__arm64__"
+          "-D__aarch64__"
+        ];
+        configureFlags = prev.configureFlags ++ [
           "--with-vmversion=5.0"
           "--with-src=src/spur64.stack"
           "--disable-cogit"
           "--without-npsqueak"
         ];
-      };
+      });
     };
+  };
 
-    "x86_64-linux" = {
-      "newspeak-cog-spur" = buildVM {
-        platformDir = "linux64x64";
-        vmName = "newspeak.cog.spur";
+  newspeak-cog-spur = buildVM {
+    vmName = "newspeak.cog.spur";
+    platforms = {
+      x86_64-linux = (final: prev: {
         scriptName = "newspeak";
```

Comment:
Instead of the platform itself being a function, is it enough for the `stdenv.mkDerivation` to take in the function, since that is what will be overridden. To put another way, can you share an example of how the current approach can facilitate overriding in a way that, say, `overrideAttrs` cannot?

Reply:
This way of doing it is quite convoluted and it could done using `overrideAttrs` on a `mkDerivation` result for each VM and platform combination to produce the same `.drv`. The reason I did it like this was to use a per-VM attrset of platforms to set `meta.platforms` for each VM package, which makes all VMs available on all platforms except broken where not supported. I also did it this way so that `mkDerivation` is called as late as possible, rather than call it for a broken VM configuration and use `overrideAttrs` later in the same file.

Reply:
I apologize for stepping away from this for a few days. Coming back to it, I think I understand the logic now like this:

![opensmalltalk-vm-1](https://github.com/NixOS/nixpkgs/assets/3806110/0e1586e8-30c1-4fa9-a939-028b417eede1)

I lean more toward something like this:

![opensmalltalkvm-2](https://github.com/NixOS/nixpkgs/assets/3806110/1efc2e04-f72f-483a-befe-2df9a8991237)

which is less flexible because it does not have the backward arrows but I feel easier to understand because of it.

But this in the end is a matter of personal taste, so if you like the current approach I don't mind.